libs/beacon/tests/dkg/trusted_dealer_test.cpp

```diff
@@ -0,0 +1,228 @@
+    , muddle{muddle::CreateMuddle("Test", muddle_certificate, network_manager, "127.0.0.1")}
+    , beacon_service{*muddle, manifest_cache, muddle_certificate, event_manager}
+    , identity{muddle_certificate->identity()}
+  {
+    network_manager.Start();
+    muddle->Start({muddle_port});
+  }
+
+  muddle::Address GetMuddleAddress() const
+  {
+    return muddle->GetAddress();
+  }
+
+  network::Uri GetHint() const
+  {
+    return fetch::network::Uri{"tcp://127.0.0.1:" + std::to_string(muddle_port)};
+  }
+};
+
+void RunTrustedDealer(uint16_t total_renewals = 4, uint32_t cabinet_size = 4,
+                      uint16_t numbers_per_aeon = 10, double threshold = 0.5)
+{
+  std::cout << "- Setup" << std::endl;
+
+  std::vector<std::unique_ptr<CabinetNode>> cabinet;
+  for (uint16_t ii = 0; ii < cabinet_size; ++ii)
+  {
+    auto port_number = static_cast<uint16_t>(10000 + ii);
+    cabinet.emplace_back(new CabinetNode{port_number, ii});
+  }
+  sleep_for(500ms);
+
+  // Connect muddles together (localhost for this example)
+  for (uint32_t ii = 0; ii < cabinet_size; ii++)
+  {
+    for (uint32_t jj = ii + 1; jj < cabinet_size; jj++)
+    {
+      cabinet[ii]->muddle->ConnectTo(cabinet[jj]->GetMuddleAddress(), cabinet[jj]->GetHint());
+    }
+  }
+
+  // wait for all the nodes to completely connect
+  std::unordered_set<uint32_t> pending_nodes;
+  for (uint32_t ii = 0; ii < cabinet_size; ++ii)
+  {
+    pending_nodes.emplace(ii);
+  }
+
+  while (!pending_nodes.empty())
+  {
+    sleep_for(100ms);
+
+    for (auto it = pending_nodes.begin(); it != pending_nodes.end();)
+    {
+      auto &muddle = *(cabinet[*it]->muddle);
+
+      if (static_cast<uint32_t>(muddle.GetNumDirectlyConnectedPeers() + 1) >= cabinet_size)
+      {
+        it = pending_nodes.erase(it);
+      }
+      else
+      {
+        ++it;
+      }
+    }
+  }
+
+  std::set<MuddleAddress> cabinet_addresses;
+  for (auto &member : cabinet)
+  {
+    cabinet_addresses.insert(member->muddle_certificate->identity().identifier());
+  }
+
+  // Attaching the cabinet logic
+  for (auto &member : cabinet)
+  {
+    member->reactor.Attach(member->beacon_service.GetMainRunnable());
+    member->reactor.Attach(member->beacon_service.GetSetupRunnable());
+  }
+
+  // Starting the beacon
+  for (auto &member : cabinet)
+  {
+    member->reactor.Start();
+  }
+
+  // Ready
+  uint64_t i = 0;
+  while (i < total_renewals)
+  {
+    std::cout << "- Scheduling round " << i << std::endl;
+    TrustedDealer dealer(cabinet_addresses,
+                         static_cast<uint32_t>(static_cast<double>(cabinet.size()) * threshold));
+    for (auto &member : cabinet)
+    {
+      member->beacon_service.StartNewCabinet(
+          cabinet_addresses, static_cast<uint32_t>(static_cast<double>(cabinet.size()) * threshold),
+          i * numbers_per_aeon, (i + 1) * numbers_per_aeon,
+          static_cast<uint64_t>(std::time(nullptr)), dealer.GetKeys(member->identity.identifier()));
```

Comment:
I would do it slightly differently to this: By creating an inherited class of beacon_service that switches out the cabinet creator for TrustedDealer which effectively has a static internally that doles out keys - could chat on Monday?

The reason I say this is that doing this allows you to do nice things like mocking, plus there is no debug-only arguments to functions when you are looking at the production code